src/rpc/zmq_pub.cpp

```diff
@@ -0,0 +1,462 @@
+
+  using chain_writer =  void(rapidjson::StringBuffer&, std::uint64_t, const cryptonote::block&);
+  using txpool_writer = void(rapidjson::StringBuffer&, epee::span<const cryptonote::txpool_event>);
+
+  template<typename F>
+  struct context
+  {
+    char const* const name;
+    F* generate_pub;
+  };
+
+  template<typename T>
+  bool operator<(const context<T>& lhs, const context<T>& rhs) noexcept
+  {
+    return std::strcmp(lhs.name, rhs.name) < 0;
+  }
+
+  template<typename T>
+  bool operator<(const context<T>& lhs, const boost::string_ref rhs) noexcept
+  {
+    return lhs.name < rhs;
+  }
+
+  struct is_valid
+  {
+    bool operator()(const cryptonote::txpool_event& event) const noexcept
+    {
+      return event.valid;
+    }
+  };
+
+  template<typename T, std::size_t N>
+  void verify_sorted(const std::array<context<T>, N>& elems, const char* name)
+  {
+    auto unsorted = std::is_sorted_until(elems.begin(), elems.end());
+    if (unsorted != elems.end())
+      throw std::logic_error{name + std::string{" array is not properly sorted, see: "} + unsorted->name};
+  }
+
+  void write_header(rapidjson::StringBuffer& buf, const boost::string_ref name)
+  {
+    PutReserve(buf, name.size());
+    for (const char ch : name)
+      PutUnsafe(buf, ch);
+    buf.Put(':');
+  }
+
+  //! \return `name:...` where `...` is JSON and `name` is directly copied (no quotes - not JSON).
+  template<typename T>
+  void json_pub(rapidjson::StringBuffer& buffer, const T value)
+  {
+    rapidjson::Writer<rapidjson::StringBuffer> dest{buffer};
+    using cryptonote::json::toJsonValue;
+    toJsonValue(dest, value);
+  }
+
+  //! Object for "minimal" block serialization
+  struct minimal_chain
+  {
+    const std::uint64_t height;
+    const cryptonote::block& block;
+  };
+
+  //! Object for "minimal" tx serialization
+  struct minimal_txpool
+  {
+    const cryptonote::transaction& tx;
+  };
+
+  void toJsonValue(rapidjson::Writer<rapidjson::StringBuffer>& dest, const minimal_chain self)
+  {
+    crypto::hash id{};
+    if (!get_block_hash(self.block, id))
+    {
+      MERROR("ZMQ/Pub failure: get_block_hash");
+      return;
+    }
+
+    dest.StartObject();
+    INSERT_INTO_JSON_OBJECT(dest, height, self.height);
+    INSERT_INTO_JSON_OBJECT(dest, prev_id, self.block.prev_id);
+    INSERT_INTO_JSON_OBJECT(dest, id, id);
+    dest.EndObject();
+  }
+
+  void toJsonValue(rapidjson::Writer<rapidjson::StringBuffer>& dest, const minimal_txpool self)
+  {
+    crypto::hash id{};
+    std::size_t blob_size = 0;
+    if (!get_transaction_hash(self.tx, id, blob_size))
+    {
+      MERROR("ZMQ/Pub failure: get_transaction_hash");
+      return;
+    }
+
+    dest.StartObject();
+    INSERT_INTO_JSON_OBJECT(dest, id, id);
+    INSERT_INTO_JSON_OBJECT(dest, blob_size, blob_size);
+    dest.EndObject();
+  }
+
+  void json_full_chain(rapidjson::StringBuffer& buf, const std::uint64_t height, const cryptonote::block& bl)
+  {
+    json_pub(buf, std::cref(bl));
```

Comment:
The JSON schema only allows one block at a time. When the daemon is synchronizing, this sends lots of individual ZMQ messages. Actually changing this will require a bit more work in the "core" code, but the schema doesn't allow any flexibility right now.

I'm thinking about making the schema:
```json
{"first_height": ..., "last_id": ..., "blocks: [ ... ]}
```
I initially decided against this for micro-optimization purposes on the receiver side - they need the equivalent of a `std::vector` which is less efficient for the more common single block case. However, this can be mitigated in several different ways (`boost::container::small_vector`, etc).

The other advantage is that the full block send doesn't contain the hash directly right now (it must be computed by duplicating what the daemon does), and getting the block height for full block requires poking into the `miner_tx`. So these fields are redundant for the full block, but still kind of helpful due to the complexity of the schema.

Any NACKs?

Reply:
I switched to arrays for `chain_main` events (future extensibility), but did not change the full block output. Minimal block events have differed slightly with array change.